libgadget/blackhole.c

```diff
@@ -1245,28 +1385,49 @@ blackhole_feedback_ngbiter(TreeWalkQueryBHFeedback * I,
             wk = density_kernel_wk(&iter->feedback_kernel, u);
-
-        const double injected_BH = I->FeedbackEnergy * mass_j * wk / I->FeedbackWeightSum;
-        /* Set a flag for star-forming particles:
-            * we want these to cool to the EEQOS via
-            * tcool rather than trelax.*/
-        if(sfreff_on_eeqos(&SPHP(other), BH_GET_PRIV(lv->tw)->a3inv)) {
-            /* We cannot atomically set a bitfield.
-             * This flag is never read in this thread loop, and we are careful not to
-             * do this with a swallowed particle (as this can race with IsGarbage being set).
-             * So lack of atomicity is (I think) not a problem.*/
-            //#pragma omp atomic write
-            P[other].BHHeated = 1;
+        
+        /* thermal feedback */
+        if(I->FeedbackEnergy > 0 && I->FdbkChannel == 0){
+            const double injected_BH = I->FeedbackEnergy * mass_j * wk / I->FeedbackWeightSum;
+            /* Set a flag for star-forming particles:
+                * we want these to cool to the EEQOS via
+                * tcool rather than trelax.*/
+            if(sfreff_on_eeqos(&SPHP(other), BH_GET_PRIV(lv->tw)->a3inv)) {
+                /* We cannot atomically set a bitfield.
+                 * This flag is never read in this thread loop, and we are careful not to
+                 * do this with a swallowed particle (as this can race with IsGarbage being set).
+                 * So lack of atomicity is (I think) not a problem.*/
+                //#pragma omp atomic write
+                P[other].BHHeated = 1;
+            }
+            const double enttou = pow(SPHP(other).Density * BH_GET_PRIV(lv->tw)->a3inv, GAMMA_MINUS1) / GAMMA_MINUS1;
+            double entold, entnew;
+            double * entptr = &(SPHP(other).Entropy);
+            #pragma omp atomic read
+            entold = *entptr;
+            do {
+                entnew = add_injected_BH_energy(entold * enttou, injected_BH, P[other].Mass) / enttou;
+                /* Swap in the new gas entropy only if the old one hasn't changed.*/
+            } while(!__atomic_compare_exchange(entptr, &entold, &entnew, 0, __ATOMIC_RELAXED, __ATOMIC_RELAXED));
+        }
+        
+        /* kinetic feedback */
+        if(I->KEFeedbackEnergy > 0 && I->FdbkChannel == 1){
+            /* Kick the gas particle*/
+            double dvel = sqrt(2 * I->KEFeedbackEnergy * wk / I->Density);
+            double dir[3];
+            get_random_dir(other, dir);
+            int j;
+            for(j = 0; j < 3; j++){
+                double velold, velnew;
+                double *velptr = &(P[other].Vel[j]);
+                #pragma omp atomic read
+                velold = *velptr;
+                do {
+                    velnew = velold + dvel*dir[j];
+                } while(!__atomic_compare_exchange(velptr, &velold, &velnew, 0, __ATOMIC_RELAXED, __ATOMIC_RELAXED));
```

Comment:
You don't actually need __atomic_compare_exchange here: it is easier and simpler to use #pragma omp atomic update.

Reply:
So something like:
for(j=0; j < 3; j++)
#pragma omp atomic update
vel[j] += (dvel * dir[j])